### DISCUSSION TITLE
Get kerning between two individual codepoints

### DISCUSSION THREAD

**Discussion by feliwir** [harfbuzz/harfbuzz#2767] [Q&A] [answered]
Hey, 

i'm using Harfbuzz for my font renderer (in addition to Freetype), since Freetype doesn't support **GPOS** kerning tables
(-> **FT_HAS_KERNING** is false). 
So i create the harfbuzz font handle from my freetype face like this (size of m_face has already been set):
```c
hb_font_t* hb_font = hb_ft_font_create (m_face, NULL);
```

My text renderer queries the horizontal kerning between two individual codepoints, so i tried using: `hb_font_get_glyph_h_kerning`, to retrieve that value with harfbuzz. However i'm not sure in which value/metric this is (i'd guess in pixels).  The function always returns 0 for the [Roboto](https://fonts.google.com/specimen/Roboto) fonts. The code looks like this:
```c
int getKerning(Unicode prev, Unicode c)
{
 return hb_font_get_glyph_h_kerning(m_internals->m_hb_font, prev, c);
}
```

Any suggestions what i'm missing here? I don't really want to shape specific strings, but just retrieve the kerning between individual codepoints. Thanks

EDIT:
As it turns out harfbuzz uses **FT_Get_Kerning** internally again, which i wanted to avoid in the first place due to the lack of **GPOS** support.... Can i disable that behavior somehow?

**Comment by khaledhosny** [comment, marked as answer]
HarfBuzz (intentionally) does not provide an API to get kerning since proper text layout involves much more than kerning. You need to use [`hb_shape()`](https://harfbuzz.github.io/harfbuzz-hb-shape.html#hb-shape) to do the text shaping, also review the [user manual](https://harfbuzz.github.io/pt01.html) for more information.

`hb_font_get_glyph_h_kerning()` only queries the legacy `kern` table (just like `FT_Get_Kerning()`) and should not generally be used.

**Reply by feliwir** [reply]
@khaledhosny well, we only support english characters / latin1 characters and kerning is really the only missing bit to get proper text support. Unfortunaly `hb_shape()` is not really a good match for our internal text rendering engine (in terms of API). Our glyphs are cached entirely inside an atlas and all glyph metrics are expected to be pretty much static.

Thanks for your information though

**Reply by feliwir** [reply]
I'll just try to shape the two individual codepoints now to get some kerning information, like this i guess:
```c
hb_buffer_t *hb_buffer;
hb_buffer = hb_buffer_create ();
hb_buffer_add(hb_buffer, prev, 0);
hb_buffer_add(hb_buffer, c, 1);
hb_buffer_guess_segment_properties (hb_buffer);

/* Shape it! */
hb_shape (m_internals->m_hb_font, hb_buffer, NULL, 0);

//TODO: calculate glyph distance

hb_buffer_destroy (hb_buffer);
  ```

**Reply by khaledhosny** [reply]
> we only support english characters / latin1 characters and kerning is really the only missing bit to get proper text support

This is a misconception. Fonts can and do all kinds of involved layout, and even monospace fonts come with ligatures name, not to mention handwriting fonts or even more subtle cases. You are doing your users a disservice not providing proper text layout.

> Our glyphs are cached entirely inside an atlas and all glyph metrics are expected to be pretty much static.

Glyph metrics are static, using `hb_shape()` does not change that (what you get out of it is glyph positions, which are not static of course, just like you do kerning but more generalized).

**Reply by khaledhosny** [reply]
> I'll just try to shape the two individual codepoints now to get some kerning information

This is worse than not doing shaping at all, since text layout is context specific. If you really really want to do it your way, then better not use HarfBuzz at all and try to find a library that does raw GPOS parsing, or implement it yourself (given your text stack limitation, you can only handle a very few number of lookup types), but it still would be the wrong thing to do.

**Reply by feliwir** [reply]
To clarify this a bit:
We used bitmap fonts till now (created with https://www.angelcode.com/products/bmfont/ ) and now are making the switch to truetype fonts. The fonts are bundled static with the application, so we won't get handwritten fonts or some other fonts all the sudden that have very complicated layouts. 

Since we preferably want to reuse most of our existing render logic we went with the approach to render the truetype fonts into a fontatlas. The only missing information that we got before with **BMFont** was the kerning pairs, that we now can't retrieve. We shape the text ourselves and replacing the existing code entirely was not really what i had in mind :-/

**Reply by khaledhosny** [reply]
You don’t need HarfBuzz then, you can extract the kerning information you want (e.g. using [FontTools](https://github.com/fonttools/fonttools/), and embed it in your application at build time.

**Comment by khaledhosny** [comment]
Also, GPOS kerning works on glyph indices not Unicode code points, and it is part of the general OpenType shaping model and is not supposed to be used in isolation.

**Reply by feliwir** [reply]
@khaledhosny well i was just mentioning **GPOS** since freetype lacks support for that. I don't really want to know about the internal representation of the kerning information, but instead just retrieve it programmatically.

**Reply by jfkthame** [reply]
It doesn't sound like you need to care about GPOS at all, if you're doing English-only text with a specific, known font where you can control what layout features it has, and just want simple pair kerning. You're explicitly not dealing with most of the complex layout behaviors that GSUB/GPOS provide and harfbuzz implements, so just stick to the legacy kern table and use freetype.

**Reply by feliwir** [reply]
@jfkthame the font contains no legacy kerning information though. Adding the legacy table with **FontForge** almost triples the font filesize

**Reply by jfkthame** [reply]
> @jfkthame the font contains no legacy kerning information though. Adding the legacy table with **FontForge** almost triples the font filesize

Presumably you could entirely strip out the OpenType tables like GSUB/GPOS that your engine doesn't support anyway, which should mitigate that.
